File: platforms/dotnet/Engine.cs
using System;
using System.Diagnostics;
using System.Text.Json;
using System.Threading.Tasks;

namespace Lpte
{
    /// <summary>
    /// LPTE — Local Profanity & Toxicity Engine for .NET.
    /// Communicates with the Python LPTE engine via subprocess IPC.
    /// </summary>
    public class Engine : IDisposable
    {
        private readonly string _languageCode;
        private readonly double _threshold;
        private readonly string _pythonPath;
        private bool _disposed;

        public Engine(string languageCode = "en", double threshold = 0.6, string pythonPath = "python3")
        {
            _languageCode = languageCode;
            _threshold = threshold;
            _pythonPath = pythonPath;
        }

        /// <summary>
        /// Analyze text for toxic content.
        /// </summary>
        public async Task<ClassificationResult> AnalyzeAsync(string text, double? threshold = null)
        {
            var t = threshold ?? _threshold;
            var script = $@"
import sys, json
sys.path.insert(0, '.')
from lpte import LpteEngine
from lpte.languages import EnglishProfile, BengaliProfile

profiles = {{'en': EnglishProfile, 'bn': BengaliProfile}}
profile = profiles.get('{_languageCode}', EnglishProfile)
engine = LpteEngine(profile)

result = engine.analyze("""{text.Replace("\"", "\\\"")}""", {t})
print(json.dumps({{
    'is_toxic': result.is_toxic,
    'severity': result.severity.value,
    'confidence': result.confidence,
    'matched_terms': result.matched_terms
}}))";

            var output = await RunPython(script);
            var raw = JsonSerializer.Deserialize<RawResult>(output);

            return new ClassificationResult
            {
                IsToxic = raw.IsToxic,
                Severity = (Severity)raw.Severity,
                Confidence = raw.Confidence,
                MatchedTerms = raw.MatchedTerms
            };
        }

        /// <summary>
        /// Quick toxicity check.
        /// </summary>
        public async Task<bool> IsToxicAsync(string text, double? threshold = null)
        {
            var result = await AnalyzeAsync(text, threshold);
            return result.IsToxic;
        }

        /// <summary>
        /// Synchronous analyze (blocks until complete).
        /// </summary>
        public ClassificationResult Analyze(string text, double? threshold = null)
        {
            return AnalyzeAsync(text, threshold).GetAwaiter().GetResult();
        }

        /// <summary>
        /// Synchronous toxicity check.
        /// </summary>
        public bool IsToxic(string text, double? threshold = null)
        {
            return IsToxicAsync(text, threshold).GetAwaiter().GetResult();
        }

        private async Task<string> RunPython(string script)
        {
            var psi = new ProcessStartInfo
            {
                FileName = _pythonPath,
                Arguments = $"-c \"{script.Replace("\"", "\\\"")}\"",
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };

            using var process = Process.Start(psi);
            var output = await process.StandardOutput.ReadToEndAsync();
            await process.WaitForExitAsync();

            if (process.ExitCode != 0)
            {
                var error = await process.StandardError.ReadToEndAsync();
                throw new LpteException($"Python error: {error}");
            }

            return output;
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                _disposed = true;
            }
        }
    }

    public enum Severity
    {
        None = 0,
        Low = 1,
        Medium = 2,
        High = 3,
        Critical = 4
    }

    public class ClassificationResult
    {
        public bool IsToxic { get; set; }
        public Severity Severity { get; set; }
        public double Confidence { get; set; }
        public string[] MatchedTerms { get; set; } = Array.Empty<string>();

        public override string ToString() =>
            $"Toxic={IsToxic}, Severity={Severity}, Confidence={Confidence:F2}";
    }

    public class LpteException : Exception
    {
        public LpteException(string message) : base(message) { }
    }

    internal class RawResult
    {
        public bool IsToxic { get; set; }
        public int Severity { get; set; }
        public double Confidence { get; set; }
        public string[] MatchedTerms { get; set; } = Array.Empty<string>();
    }
}

File: platforms/php/Engine.php
<?php
/**
 * LPTE — PHP wrapper for the Local Profanity & Toxicity Engine.
 *
 * Communicates with the Python LPTE engine via proc_open() subprocess IPC.
 *
 * Usage:
 *   $engine = new LpteEngine('en');
 *   $result = $engine->analyze('some text');
 *   if ($result['is_toxic']) {
 *       echo "Toxic: {$result['severity']}\n";
 *   }
 */

namespace Lpte;

class Engine
{
    private string $languageCode;
    private float $threshold;
    private string $pythonPath;

    public function __construct(string $languageCode = 'en', float $threshold = 0.6, string $pythonPath = 'python3')
    {
        $this->languageCode = $languageCode;
        $this->threshold = $threshold;
        $this->pythonPath = $pythonPath;
    }

    /**
     * Analyze text for toxic content.
     *
     * @param string $text Input text
     * @param float|null $threshold Classification threshold (0.0-1.0)
     * @return array{is_toxic: bool, severity: string, confidence: float, matched_terms: string[]}
     */
    public function analyze(string $text, ?float $threshold = null): array
    {
        $t = $threshold ?? $this->threshold;
        $escapedText = addslashes($text);

        $script = <<<PYTHON
import sys, json
sys.path.insert(0, '.')
from lpte import LpteEngine
from lpte.languages import EnglishProfile, BengaliProfile

profiles = {'en': EnglishProfile, 'bn': BengaliProfile}
profile = profiles.get('{$this->languageCode}', EnglishProfile)
engine = LpteEngine(profile)

result = engine.analyze("""{$escapedText}""", {$t})
print(json.dumps({
    'is_toxic': result.is_toxic,
    'severity': result.severity.name,
    'confidence': result.confidence,
    'matched_terms': result.matched_terms
}))
PYTHON;

        $output = $this->runPython($script);
        return json_decode($output, true) ?? [
            'is_toxic' => false,
            'severity' => 'NONE',
            'confidence' => 0.0,
            'matched_terms' => [],
        ];
    }

    /**
     * Quick toxicity check.
     */
    public function isToxic(string $text, ?float $threshold = null): bool
    {
        $result = $this->analyze($text, $threshold);
        return $result['is_toxic'] ?? false;
    }

    /**
     * Sanitize text by masking toxic segments.
     */
    public function sanitize(string $text, string $mask = '*'): string
    {
        $escapedText = addslashes($text);
        $escapedMask = addslashes($mask);

        $script = <<<PYTHON
import sys, json
sys.path.insert(0, '.')
from lpte import LpteEngine
from lpte.languages import EnglishProfile

engine = LpteEngine(EnglishProfile)
result = engine.sanitize("""{$escapedText}""", "{$escapedMask}")
print(json.dumps({'result': result}))
PYTHON;

        $output = $this->runPython($script);
        $decoded = json_decode($output, true);
        return $decoded['result'] ?? $text;
    }

    private function runPython(string $script): string
    {
        $descriptors = [
            0 => ['pipe', 'r'],  // stdin
            1 => ['pipe', 'w'],  // stdout
            2 => ['pipe', 'w'],  // stderr
        ];

        $process = proc_open(
            "{$this->pythonPath} -c " . escapeshellarg($script),
            $descriptors,
            $pipes
        );

        if (!is_resource($process)) {
            throw new \RuntimeException('Failed to start Python process');
        }

        fclose($pipes[0]);
        $output = stream_get_contents($pipes[1]);
        fclose($pipes[1]);
        fclose($pipes[2]);

        $exitCode = proc_close($process);
        if ($exitCode !== 0) {
            throw new \RuntimeException("Python process failed with exit code {$exitCode}");
        }

        return trim($output);
    }
}
